src/mongo_odm/query_builder.hpp

```diff
@@ -0,0 +1,356 @@
+// Copyright 2016 MongoDB Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <mongo_odm/config/prelude.hpp>
+
+#include <cstddef>
+#include <stdexcept>
+#include <tuple>
+#include <type_traits>
+
+#include <bsoncxx/builder/core.hpp>
+#include <bsoncxx/view_or_value.hpp>
+
+#include <mongo_odm/nvp.hpp>
+
+namespace mongo_odm {
+MONGO_ODM_INLINE_NAMESPACE_BEGIN
+
+// Forward declarations
+template <typename Base, typename T>
+class NotExpr;
+
+/**
+ * Represents a binary comparison expression between a key and a value. e.g. (User.age > 21).
+ */
+template <typename Base, typename T>
+class ComparisonExpr {
+   public:
+    /**
+     * Constructs a query expression for the given key, value, and comparison type
+     * @param  nvp           A name-value pair corresponding to a key in a document
+     * @param  field         The value that the key is being compared to.
+     * @param  selector_type The type of comparison operator, such at gt (>) or ne (!=).
+     */
+    constexpr ComparisonExpr(const Nvp<Base, T> &nvp, T field, const char *selector_type)
+        : _nvp(nvp), _field(std::move(field)), _selector_type(selector_type) {
+    }
+    /**
+     * Appends this expression to a BSON core builder, as a key-value pair of the form
+     * "key: {$cmp: val}", where $cmp is some comparison operator.
+     * @param builder A BSON core builder
+     */
+    void append_to_bson(bsoncxx::builder::core &builder) const {
+        builder.key_view(_nvp.name);
+        builder.open_document();
+        builder.key_view(_selector_type);
+        builder.append(_field);
+        builder.close_document();
+    }
+
+    /**
+     * Converts the expression to a BSON filter for a query.
+     * The resulting BSON is of the form "{key: {$cmp: val}}".
+     */
+    operator bsoncxx::document::view_or_value() const {
```

Comment:
Since the `mongocxx::collection` methods accept `view_or_value`, I don't think it'd be possible to implicitly convert to expressions to `value`, and then have that be implicitly converted again to `view_or_value`. i.e. The code `col.find_one(ODM_KEY(Foo::x) == 1)` would have to be `col.find_one(view_or_value(ODM_KEY(Foo::x) == 1))`, which I think would be unnecessarily verbose.

At least, when I use `operator bsoncxx::document::value() const`, the tests fail to compile unless I explicitly case expressions to `view_or_value`.